zap/src/output/client.rs

```diff
@@ -1,260 +1,363 @@
-use std::fmt::Display;
-
 use crate::{
-	irgen::{gen_des, gen_ser},
-	parser::{EvCall, EvSource, EvType, File},
+	irgen::{gen_des, gen_ser, Stmt},
+	parser::{EvCall, EvDecl, EvSource, EvType, File, TyDecl},
 	util::casing,
 };
 
-struct ClientFile<'a>(&'a File);
+use super::Output;
+
+struct ClientOutput<'a> {
+	file: &'a File,
+	buff: String,
+	tabs: u32,
+}
+
+impl<'a> Output for ClientOutput<'a> {
+	fn push(&mut self, s: &str) {
+		self.buff.push_str(s);
+	}
+
+	fn indent(&mut self) {
+		self.tabs += 1;
+	}
+
+	fn dedent(&mut self) {
+		self.tabs -= 1;
+	}
+
+	fn push_indent(&mut self) {
+		for _ in 0..self.tabs {
+			self.push("\t");
+		}
+	}
+}
+
+impl<'a> ClientOutput<'a> {
+	pub fn new(file: &'a File) -> Self {
+		Self {
+			file,
+			buff: String::new(),
+			tabs: 0,
+		}
+	}
+
+	fn push_tydecl(&mut self, tydecl: &TyDecl) {
+		let name = &tydecl.name;
+		let ty = &tydecl.ty;
+
+		self.push_line(&format!("export type {name} = {ty}"));
+
+		self.push_line(&format!("function types.write_{name}(value: {name})"));
+		self.push_stmts(&gen_ser(ty, "value".into(), self.file.write_checks));
+		self.push_line("end");
+
+		self.push_line(&format!("function types.read_{name}()"));
+		self.push_line("local value;");
+		self.push_stmts(&gen_des(ty, "value".into(), false));
+		self.push_line("return value");
+		self.push_line("end");
+	}
+
+	fn push_tydecls(&mut self) {
+		for tydecl in self.file.ty_decls.iter() {
+			self.push_tydecl(tydecl);
+		}
+	}
+
+	fn push_reliable_header(&mut self) {
+		self.push_line("reliable.OnClientEvent:Connect(function(buff, inst)");
+		self.indent();
+		self.push_line("incoming_buff = buff");
+		self.push_line("incoming_inst = inst");
+		self.push_line("incoming_read = 0");
 
-impl<'a> Display for ClientFile<'a> {
-	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-		let file = self.0;
+		self.push_line("local len = buffer.len(buff)");
+		self.push_line("while incoming_read < len do");
 
-		if file.ev_decls.is_empty() {
-			return Ok(());
+		self.indent();
+
+		self.push_line(&format!(
+			"local id = buffer.read{}(buff, read({}))",
+			self.file.event_id_ty(),
+			self.file.event_id_ty().size()
+		));
+	}
+	fn push_reliable_callback(&mut self, first: bool, ev: &EvDecl, id: usize) {
+		self.push_indent();
+
+		if first {
+			self.push("if ");
+		} else {
+			self.push("elseif ");
 		}
 
-		writeln!(f, "--!native")?;
-		writeln!(f, "--!optimize 2")?;
-		writeln!(f, "-- Generated by zap v{}", env!("CARGO_PKG_VERSION"))?;
+		self.push(&format!("id == {id} then"));
+		self.push("\n");
```

Comment:
Shouldn't this just be:
```suggestion
		self.push_line(&format!("id == {id} then"));
```

Reply:
`push_line` also pushes tabs, so we can't do this. We could move `\n` to the same push though.

Reply:
Let's keep them explicitly seperate and add a comment as to why we aren't using `push_line`.